custom_components/elasticsearch/es_publish_pipeline.py

```diff
@@ -377,19 +550,167 @@ def __init__(self, hass: HomeAssistant, log: Logger = BASE_LOGGER) -> None:
+
+                new_value = value
+
+                if isinstance(value, dict):
+                    new_value = json.dumps(value)
+                elif isinstance(value, set):
+                    new_value = list(value)
+                elif (
+                    isinstance(value, list | tuple)
+                    and len(value) > 0
+                    and isinstance(value[0], tuple | dict | set | list)
+                ):
+                    new_value = json.dumps(value)
+
+                attributes[new_key] = new_value
+
+            return attributes
+
+        def state_to_coerced_value(self, state: State) -> dict:
+            """Coerce the state value into a dictionary of possible types."""
+            value = state.state
+
+            success, result = self.try_state_as_boolean(state)
+            if success and result is not None:
+                return {"boolean": result}
+
+            success, result = self.try_state_as_number(state)
+            if success and result is not None:
+                return {"float": result}
+
+            success, result = self.try_state_as_datetime(state)
+            if success and result is not None:
+                return {
+                    "datetime": result.isoformat(),
+                    "date": result.date().isoformat(),
+                    "time": result.time().isoformat(),
+                }
+
+            return {"string": value}
+
+        # Static converter helpers
+
         @staticmethod
         @lru_cache(maxsize=128)
-        def _sanitize_domain(domain: str) -> str:
+        def sanitize_domain(domain: str) -> str:
             """Sanitize the domain name."""
             # Only allow alphanumeric characters a-z 0-9 and underscores
             return re.sub(r"[^a-z0-9_]", "", domain.lower())[0:128]
 
+        @staticmethod
+        def domain_to_datastream(domain: str) -> dict:
+            """Convert the state into a datastream."""
+            return {
+                "datastream.type": DATASTREAM_TYPE,
+                "datastream.dataset": DATASTREAM_DATASET_PREFIX
+                + "."
+                + Pipeline.Formatter.sanitize_domain(domain),
+                "datastream.namespace": DATASTREAM_NAMESPACE,
+            }
+
+        def filter_attribute(self, entity_id, key, value) -> bool:
+            """Filter out attributes we don't want to publish."""
+            msg: str | None = None
+
+            if key in SKIP_ATTRIBUTES:
+                msg = f"Attribute {key} is a skippable attribute. It has value {value} and is from entity {entity_id}."
+
+            elif not isinstance(key, ALLOWED_ATTRIBUTE_KEY_TYPES):
+                msg = f"Attribute {key} has a disallowed key type {type(key)}. It has value {value} from entity {entity_id}."
+
+            elif not isinstance(value, ALLOWED_ATTRIBUTE_VALUE_TYPES):
+                msg = f"Attribute {key} has a disallowed value type {type(value)}. It has value {value} from entity {entity_id}."
+
```

Comment:
I wish there was a `verbose` or `silly` log level, as I find these entries to be quite noisy when debug logging is enabled.
Do you think we get enough benefit from these messages to include in debug logs?

Reply:
Perhaps I can hide them behind a "very verbose" option that we can toggle by hand editing the config entry?

Reply:
I'm not opposed to that approach. It's additional complexity, but I'd prefer that over having to filter these out any time I wanted to debug something else with the integration.

Reply:
I would leave it disabled by default -- I just want to retain the ability to troubleshoot, "Why is my entity not being published" in a users environment using logs if possible. will include in next update to PR